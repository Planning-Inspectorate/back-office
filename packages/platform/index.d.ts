import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
