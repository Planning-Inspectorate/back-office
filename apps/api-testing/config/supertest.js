// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
