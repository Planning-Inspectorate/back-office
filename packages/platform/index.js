export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
