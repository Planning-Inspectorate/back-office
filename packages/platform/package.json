{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
