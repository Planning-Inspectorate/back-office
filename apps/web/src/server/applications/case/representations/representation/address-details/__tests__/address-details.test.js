import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { createTestEnvironment } from '../../../../../../../../testing/index.js';
import { representationFixture } from '../../../__fixtures__/representations.fixture.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);

const mockCaseReference = { title: 'mock title', status: 'in test', reference: 'mock reference' };
const nocks = () => {
	nock('http://test/').get('/applications/1').reply(200, mockCaseReference);
	nock('http://test/').get('/applications/1/representations/1').reply(200, representationFixture);
};

describe('Representation address details page', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	afterAll(() => {
		nock.cleanAll();
	});

	const baseUrl =
		'/applications-service/case/1/relevant-representations/address-details?repId=1&repType=represented';
	const lookupUrl = `${baseUrl}&stage=lookup`;
	const findUrl = `${baseUrl}&stage=find`;
	const enterUrl = `${baseUrl}&stage=enter`;

	describe('GET /applications-service/case/1/relevant-representations/address-details', () => {
		beforeEach(async () => {
			nocks();
			await request.get('/applications-service/');
		});

		it('should render the page and have the lookup fields', async () => {
			const response = await request.get(lookupUrl);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
			expect(element.innerHTML).toContain('Postcode');
		});

		it('should render the page and have the find fields', async () => {
			const response = await request.get(findUrl);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
			expect(element.innerHTML).toContain('Address');
		});

		it('should render the page and have the enter fields', async () => {
			const response = await request.get(enterUrl);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
			expect(element.innerHTML).toContain('Address line 1');
			expect(element.innerHTML).toContain('Address line 2');
			expect(element.innerHTML).toContain('Town or city');
			expect(element.innerHTML).toContain('Postcode');
			expect(element.innerHTML).toContain('Country');
		});
	});

	describe('POST /applications-service/case/1/relevant-representations/address-details', () => {
		beforeEach(async () => {
			nocks();
			await request.get('/applications-service/');
		});
		describe('Field validation:', () => {
			it('should show the lookup page validation errors', async () => {
				const response = await request.post(baseUrl).send({ stage: 'lookup' });
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain('Enter your postcode');
			});

			it('should show the find page validation errors', async () => {
				const response = await request.post(baseUrl).send({ stage: 'find' });
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain('Choose an address from the list');
			});

			it('should show the enter page validation errors', async () => {
				const response = await request.post(baseUrl).send({ stage: 'enter' });
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain('Enter address line 1');
				expect(element.innerHTML).toContain('Enter a postcode');
				expect(element.innerHTML).toContain('Enter a country');
			});
		});
	});
});
