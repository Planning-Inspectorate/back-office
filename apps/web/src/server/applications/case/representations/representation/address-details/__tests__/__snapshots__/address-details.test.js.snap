// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Representation address details page GET /applications-service/case/1/relevant-representations/address-details should render the page and have the enter fields 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <p class=\\"govuk-body\\"><strong>Project name:</strong> mock title</p><span class=\\"govuk-caption-l\\">Add a representation</span>
    <h1     class=\\"govuk-heading-l\\">Address details</h1>
        <form method=\\"post\\" novalidate>
            <input type=\\"hidden\\" name='stage' value=\\"enter\\">
            <input type=\\"hidden\\" name='setPostcode' value=\\"\\">
            <fieldset class=\\"govuk-fieldset govuk-!-width-two-thirds\\">
                <legend class=\\"govuk-fieldset__legend govuk-visually-hidden\\">What is your address?</legend>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"addressLine1\\">Address line 1</label>
                    <input class=\\"govuk-input\\" id=\\"addressLine1\\" name=\\"addressLine1\\"
                    type=\\"text\\" value=\\"96 The Avenue\\" autocomplete=\\"address-line1\\">
                </div>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"addressLine2\\">Address line 2 (optional)</label>
                    <input class=\\"govuk-input\\" id=\\"addressLine2\\"
                    name=\\"addressLine2\\" type=\\"text\\" value=\\"Maidstone\\" autocomplete=\\"address-line2\\">
                </div>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"town\\">Town or city (optional)</label>
                    <input class=\\"govuk-input govuk-!-width-two-thirds\\"
                    id=\\"town\\" name=\\"town\\" type=\\"text\\" autocomplete=\\"address-level2\\">
                </div>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"postcode\\">Postcode</label>
                    <input class=\\"govuk-input govuk-input--width-10\\" id=\\"postcode\\"
                    name=\\"postcode\\" type=\\"text\\" value=\\"MD21 5XY\\" autocomplete=\\"postal-code\\">
                </div>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"country\\">Country</label>
                    <input class=\\"govuk-input govuk-!-width-two-thirds\\" id=\\"country\\"
                    name=\\"country\\" type=\\"text\\">
                </div>
            </fieldset>
            <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
        </form>
</main>"
`;

exports[`Representation address details page GET /applications-service/case/1/relevant-representations/address-details should render the page and have the find fields 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <p class=\\"govuk-body\\"><strong>Project name:</strong> mock title</p><span class=\\"govuk-caption-l\\">Add a representation</span>
    <h1     class=\\"govuk-heading-l\\">Address details</h1>
        <form method=\\"post\\" novalidate>
            <input type=\\"hidden\\" name='stage' value=\\"find\\">
            <input type=\\"hidden\\" name='setPostcode' value=\\"\\">
            <div class=\\"govuk-!-width-one-half govuk-!-margin-bottom-4\\">
                <p class=\\"govuk-label govuk-!-margin-top-0\\">Postcode</p>
                <div class=\\"govuk-grid-row\\">
                    <div class=\\"govuk-grid-column-one-half\\">
                        <p class=\\"govuk-body\\"><strong></strong>
                        </p>
                    </div>
                    <div class=\\"govuk-grid-column-one-half\\"><a class=\\"govuk-link govuk-body govuk-!-display-inline-block\\" href=\\"?repId=1&amp;repType=represented&amp;stage=lookup\\"> Change</a>
                    </div>
                </div>
            </div>
            <fieldset class=\\"govuk-fieldset govuk-!-width-two-thirds\\">
                <legend class=\\"govuk-fieldset__legend govuk-visually-hidden\\">Select your address</legend>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label\\" for=\\"address\\">Address</label>
                    <select class=\\"govuk-select\\" id=\\"address\\" name=\\"address\\">
                        <option value=\\"\\" selected disabled hidden=\\"true\\">0 addresses found</option>
                    </select>
                </div>
            </fieldset>
            <div class=\\"govuk-!-margin-bottom-6\\"><a class='govuk-link govuk-body govuk-!-display-inline-block govuk-!-margin-0'
                href=\\"?repId=1&amp;repType=represented&amp;stage=enter\\"> I can't find address on the list</a>
            </div>
            <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
        </form>
</main>"
`;

exports[`Representation address details page GET /applications-service/case/1/relevant-representations/address-details should render the page and have the lookup fields 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <p class=\\"govuk-body\\"><strong>Project name:</strong> mock title</p><span class=\\"govuk-caption-l\\">Add a representation</span>
    <h1     class=\\"govuk-heading-l\\">Address details</h1>
        <form method=\\"post\\" novalidate>
            <input type=\\"hidden\\" name='stage' value=\\"lookup\\">
            <input type=\\"hidden\\" name='setPostcode' value=\\"\\">
            <fieldset class=\\"govuk-fieldset\\">
                <legend class=\\"govuk-fieldset__legend govuk-visually-hidden\\">Enter your postcode</legend>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label\\" for=\\"lookupPostcode\\">Postcode</label>
                    <input class=\\"govuk-input govuk-!-width-one-third\\" id=\\"lookupPostcode\\"
                    name=\\"lookupPostcode\\" type=\\"text\\">
                </div>
            </fieldset>
            <div class=\\"govuk-!-margin-bottom-6\\"><a class='govuk-link govuk-body govuk-!-display-inline-block govuk-!-margin-0'
                href=\\"?repId=1&amp;repType=represented&amp;stage=enter\\"> Don't have a UK postcode?</a>
            </div>
            <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Find address</button>
        </form>
</main>"
`;

exports[`Representation address details page POST /applications-service/case/1/relevant-representations/address-details Field validation: should show the enter page validation errors 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <div class=\\"govuk-error-summary\\" data-module=\\"govuk-error-summary\\">
                <div role=\\"alert\\">
                    <h2 class=\\"govuk-error-summary__title\\"> There is a problem</h2>
                    <div class=\\"govuk-error-summary__body\\">
                        <ul class=\\"govuk-list govuk-error-summary__list\\">
                            <li><a href=\\"#addressLine1\\">Enter address line 1</a>
                            </li>
                            <li><a href=\\"#postcode\\">Enter a postcode</a>
                            </li>
                            <li><a href=\\"#country\\">Enter a country</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <p class=\\"govuk-body\\"><strong>Project name:</strong> mock title</p><span class=\\"govuk-caption-l\\">Add a representation</span>
    <h1     class=\\"govuk-heading-l\\">Address details</h1>
        <form method=\\"post\\" novalidate>
            <input type=\\"hidden\\" name='stage' value=\\"enter\\">
            <input type=\\"hidden\\" name='setPostcode' value=\\"\\">
            <fieldset class=\\"govuk-fieldset govuk-!-width-two-thirds\\">
                <legend class=\\"govuk-fieldset__legend govuk-visually-hidden\\">What is your address?</legend>
                <div class=\\"govuk-form-group govuk-form-group--error\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"addressLine1\\">Address line 1</label>
                    <p id=\\"addressLine1-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> Enter address line 1</p>
                    <input                     class=\\"govuk-input govuk-input--error\\" id=\\"addressLine1\\" name=\\"addressLine1\\"
                    type=\\"text\\" aria-describedby=\\"addressLine1-error\\" autocomplete=\\"address-line1\\">
                </div>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"addressLine2\\">Address line 2 (optional)</label>
                    <input class=\\"govuk-input\\" id=\\"addressLine2\\"
                    name=\\"addressLine2\\" type=\\"text\\" autocomplete=\\"address-line2\\">
                </div>
                <div class=\\"govuk-form-group\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"town\\">Town or city (optional)</label>
                    <input class=\\"govuk-input govuk-!-width-two-thirds\\"
                    id=\\"town\\" name=\\"town\\" type=\\"text\\" autocomplete=\\"address-level2\\">
                </div>
                <div class=\\"govuk-form-group govuk-form-group--error\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"postcode\\">Postcode</label>
                    <p id=\\"postcode-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> Enter a postcode</p>
                    <input                     class=\\"govuk-input govuk-input--width-10 govuk-input--error\\" id=\\"postcode\\"
                    name=\\"postcode\\" type=\\"text\\" aria-describedby=\\"postcode-error\\" autocomplete=\\"postal-code\\">
                </div>
                <div class=\\"govuk-form-group govuk-form-group--error\\">
                    <label class=\\"govuk-label govuk-!-font-weight-bold\\" for=\\"country\\">Country</label>
                    <p id=\\"country-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> Enter a country</p>
                    <input                     class=\\"govuk-input govuk-!-width-two-thirds govuk-input--error\\" id=\\"country\\"
                    name=\\"country\\" type=\\"text\\" aria-describedby=\\"country-error\\">
                </div>
            </fieldset>
            <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
        </form>
</main>"
`;

exports[`Representation address details page POST /applications-service/case/1/relevant-representations/address-details Field validation: should show the find page validation errors 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <div class=\\"govuk-error-summary\\" data-module=\\"govuk-error-summary\\">
                <div role=\\"alert\\">
                    <h2 class=\\"govuk-error-summary__title\\"> There is a problem</h2>
                    <div class=\\"govuk-error-summary__body\\">
                        <ul class=\\"govuk-list govuk-error-summary__list\\">
                            <li><a href=\\"#address\\">Choose an address from the list</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <p class=\\"govuk-body\\"><strong>Project name:</strong> mock title</p><span class=\\"govuk-caption-l\\">Add a representation</span>
    <h1     class=\\"govuk-heading-l\\">Address details</h1>
        <form method=\\"post\\" novalidate>
            <input type=\\"hidden\\" name='stage' value=\\"find\\">
            <input type=\\"hidden\\" name='setPostcode' value=\\"\\">
            <div class=\\"govuk-!-width-one-half govuk-!-margin-bottom-4\\">
                <p class=\\"govuk-label govuk-!-margin-top-0\\">Postcode</p>
                <div class=\\"govuk-grid-row\\">
                    <div class=\\"govuk-grid-column-one-half\\">
                        <p class=\\"govuk-body\\"><strong></strong>
                        </p>
                    </div>
                    <div class=\\"govuk-grid-column-one-half\\"><a class=\\"govuk-link govuk-body govuk-!-display-inline-block\\" href=\\"?repId=1&amp;repType=represented&amp;stage=lookup\\"> Change</a>
                    </div>
                </div>
            </div>
            <fieldset class=\\"govuk-fieldset govuk-!-width-two-thirds\\">
                <legend class=\\"govuk-fieldset__legend govuk-visually-hidden\\">Select your address</legend>
                <div class=\\"govuk-form-group govuk-form-group--error\\">
                    <label class=\\"govuk-label\\" for=\\"address\\">Address</label>
                    <p id=\\"address-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> Choose an address from
                        the list</p>
                    <select class=\\"govuk-select govuk-select--error\\" id=\\"address\\"
                    name=\\"address\\" aria-describedby=\\"address-error\\">
                        <option value=\\"\\" selected disabled hidden=\\"true\\">0 addresses found</option>
                    </select>
                </div>
            </fieldset>
            <div class=\\"govuk-!-margin-bottom-6\\"><a class='govuk-link govuk-body govuk-!-display-inline-block govuk-!-margin-0'
                href=\\"?repId=1&amp;repType=represented&amp;stage=enter\\"> I can't find address on the list</a>
            </div>
            <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
        </form>
</main>"
`;

exports[`Representation address details page POST /applications-service/case/1/relevant-representations/address-details Field validation: should show the lookup page validation errors 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <div class=\\"govuk-error-summary\\" data-module=\\"govuk-error-summary\\">
                <div role=\\"alert\\">
                    <h2 class=\\"govuk-error-summary__title\\"> There is a problem</h2>
                    <div class=\\"govuk-error-summary__body\\">
                        <ul class=\\"govuk-list govuk-error-summary__list\\">
                            <li><a href=\\"#lookupPostcode\\">Enter a postcode</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <p class=\\"govuk-body\\"><strong>Project name:</strong> mock title</p><span class=\\"govuk-caption-l\\">Add a representation</span>
    <h1     class=\\"govuk-heading-l\\">Address details</h1>
        <form method=\\"post\\" novalidate>
            <input type=\\"hidden\\" name='stage' value=\\"lookup\\">
            <input type=\\"hidden\\" name='setPostcode' value=\\"\\">
            <fieldset class=\\"govuk-fieldset\\">
                <legend class=\\"govuk-fieldset__legend govuk-visually-hidden\\">Enter your postcode</legend>
                <div class=\\"govuk-form-group govuk-form-group--error\\">
                    <label class=\\"govuk-label\\" for=\\"lookupPostcode\\">Postcode</label>
                    <p id=\\"lookupPostcode-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> Enter a postcode</p>
                    <input                     class=\\"govuk-input govuk-!-width-one-third govuk-input--error\\" id=\\"lookupPostcode\\"
                    name=\\"lookupPostcode\\" type=\\"text\\" aria-describedby=\\"lookupPostcode-error\\">
                </div>
            </fieldset>
            <div class=\\"govuk-!-margin-bottom-6\\"><a class='govuk-link govuk-body govuk-!-display-inline-block govuk-!-margin-0'
                href=\\"?repId=1&amp;repType=represented&amp;stage=enter\\"> Don't have a UK postcode?</a>
            </div>
            <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Find address</button>
        </form>
</main>"
`;
